sim/src/kernels/KP07_kernel.opencl

```diff
@@ -207,73 +237,176 @@ void init_H_with_garbage(__local float Hi[block_height+4][block_width+4],
 			 __local float RHy[block_height+4][block_width+4] ) {
 
     if (get_local_id(0) == 0 && get_local_id(1) == 0) {
-	for (int j = 0; j < block_height+4; j++) {
-	    for (int i = 0; i < block_width+4; i++) {
-		Hi[j][i]  = 99.0f; //0.1*get_global_id(0);
-		RHx[j][i] = 99.0f; //0.1*get_global_id(0);
-		RHy[j][i] = 99.0f; //0.1*get_global_id(0);
-	    }
-	}
+        for (int j = 0; j < block_height+4; j++) {
+            for (int i = 0; i < block_width+4; i++) {
+            Hi[j][i]  = -99.0f; //0.1*get_global_id(0);
+            RHx[j][i] = -99.0f; //0.1*get_global_id(0);
+            RHy[j][i] = -99.0f; //0.1*get_global_id(0);
+            }
+        }
     }
 }
 
+float findMaxDt(float per_thread_dt, int tid, __local float* sdata) {
+    const int nthreads = block_width*block_height;
+    
+    //Store per-thread dt in local mem
+    sdata[tid] = per_thread_dt;
+    barrier(CLK_LOCAL_MEM_FENCE);
+    
+    //Local memory reduction to find min
+    //First reduce to 64 elements with synchronization
+    /********************************************************************
+      * WARNING: This assumes that we have at least 64 threads to work
+      *******************************************************************/
+    if (nthreads >= 512) {
+        if (tid < 512 && (tid+512) < nthreads) {
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 512]);
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    if (nthreads >= 256) {
+        if (tid < 256 && (tid+256) < nthreads) {
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 256]);
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    if (nthreads >= 128) {
+        if (tid < 128 && (tid+128) < nthreads) {
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 128]);
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    if (nthreads >= 64) {
+        if (tid < 64 && (tid+64) < nthreads) {
+            sdata[tid] = fmin(sdata[tid], sdata[tid + 64]);
+        }
+        barrier(CLK_LOCAL_MEM_FENCE);
+    }
+    
+#ifdef NVIDIA
```

Comment:
Note: This macro is not really well defined... Right now it's hardcoded in the common.opencl file. So for an Intel CPU e.g., you would need to remove the definition from there to get the right code path